### PR TITLE
Allow client to provide the OpenAIClient and APIM example

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context
  1. Why is this change required? Demonstrates how to use SK together with Azure API Management
  2. What problem does it solve? Allows a custom OpenAIClient to be created by client code and thereby have full access to `OpenAIClientOptions`
  3. What scenario does it contribute to? Support for customer who want to protect their Azure OpenAI API keys 
  4. If it fixes an open issue, please link to the issue here.

### Description
Allowing the OpenAIClient to be created by the client gives them full access to the `OpenAIClientOptions`.
Consider the case where a developer is using APIM to protect their API keys and starts to see a problem. 
If you run the associated sample with a bad subscription key the trace in will include the following:

```
      Headers:
      ErrorSource: authorization
      ErrorReason: SubscriptionKeyInvalid
      ErrorMessage: Access denied due to invalid subscription key. Make sure to provide a valid key for an active subscription.
      ErrorSection: backend
      ErrorStatusCode: 401
```

Without the `Diagnostics` seeting in `OpenAIClientOptions` the trace in will include the following:

```
      Headers:
      ErrorSource: REDACTED
      ErrorReason: REDACTED
      ErrorMessage: REDACTED
      ErrorSection: REDACTED
      ErrorStatusCode: REDACTED
```

### Contribution Checklist
- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
From PR https://github.com/microsoft/semantic-kernel/pull/1520

@shawncal 
I see why you've done this, but it's breaking my brain.... We're passing in an OpenAIClient (from Azure SDK?) to an [AzureOpenAIClientBase](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/AzureOpenAIClientBase.cs), which is also a [ClientBase](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs), but not an [OpenAIClientBase](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/OpenAIClientBase.cs), or the other [OpenAIClientBase](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.AI.OpenAI/CustomClient/OpenAIClientBase.cs).

We really should take a closer look here and clean this up.

For this PR: Personally, I'd rather not expose the actual Azure SDK classes in our public API surface. If we were to pass through configuration or specific properties instead of OpenAIClient objects to meet the APIM needs, what would those be?

@shawncal The OpenAIClient uses a class called OpenAIClientOptions so we would have to create out own abstraction for this which also feels wrong.
Why is not exposing Azure SDK classes in the public API important?

**Review by dmytrostruk — COMMENTED** [review]
@markwallace-microsoft Thank you for this PR and APIM example!

Regarding `OpenAIClient` injection - I'm not sure about that.

For APIM, it's still should be possible to use it as it is right now. In `AzureOpenAIClientBase` there is constructor which accepts `TokenCredential credential` and `string endpoint`, so in theory it should be possible to pass APIM endpoint and access token in current implementation.

As for full access to `OpenAIClientOptions` - we should think whether it provides some benefits.
I checked its properties:
- `Transport` - wrapper around `HttpClient` and we already use it.
- `Diagnostics` - this one for troubleshooting which we don't have at the moment.
- `Retry` and `RetryPolicy` - for retry strategies, but I believe we already have a possibility to configure retry strategy through Kernel instance, or it can be done on `HttpClient` level as well.

So, it looks like only `Diagnostics` property is missing, but again, in case of invalid subscription key, our current implementation should throw appropriate exception with error/status code and exception message. For better observability we can also configure `OpenAIClient` on our side with default `Diagnostics` values to provide more transparency for troubleshooting. We can also expose `List<string>` in Azure/OpenAI client constructors as optional parameter for `LoggedHeaderNames` property, but it will be abstract enough.
 
As per current implementation, we use plain .NET types to initialize Azure/OpenAI clients, like `string` endpoint, `string` key, `HttpClient`, `ILogger` etc. In case if we decide to replace Azure SDK with something else - it should not be a problem. 
But if we add Azure SDK as dependency in constructor, it will be harder to decouple it later.

File: dotnet/samples/KernelSyntaxExamples/Example52_ApimAuth.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+// ReSharper disable once InconsistentNaming
+public static class Example52_ApimAuth
+{
+    public static async Task RunAsync()
+    {
+        // Azure API Management details
+        // For more information see 'Protect your Azure OpenAI API keys with Azure API Management' here: https://learn.microsoft.com/en-us/semantic-kernel/deploy/
+        var apimUri = new Uri("https://apim...azure-api.net/");
+        var subscriptionKey = "<Add your subscription key here>";
```
[the remark this answers is not in the document]
Comment (shawncal) [reply dotnet/samples/KernelSyntaxExamples/Example52_ApimAuth.cs:23]: I've been updating this everywhere in KernelSyntaxExamples. See open PR here: https://github.com/microsoft/semantic-kernel/pull/1832

File: .gitignore [hunk 2]
```diff
@@ -476,4 +476,7 @@ playwright-report/
 
 # Static Web App deployment config
 swa-cli.config.json
-**/copilot-chat-app/webapp/build
\ No newline at end of file
+**/copilot-chat-app/webapp/build
+
+# Semantic Kernel Tools
+/.semantic-kernel
```
Comment (shawncal) [inline .gitignore:482]: what's this for?

**Review by shawncal — DISMISSED** [review]

**Comment by markwallace-microsoft** [comment]
> @markwallace-microsoft Thank you for this PR and APIM example!
> 
> Regarding `OpenAIClient` injection - I'm not sure about that.
> 
> For APIM, it's still should be possible to use it as it is right now. In `AzureOpenAIClientBase` there is constructor which accepts `TokenCredential credential` and `string endpoint`, so in theory it should be possible to pass APIM endpoint and access token in current implementation.
> 
> As for full access to `OpenAIClientOptions` - we should think whether it provides some benefits. I checked its properties:
> 
> * `Transport` - wrapper around `HttpClient` and we already use it.
> * `Diagnostics` - this one for troubleshooting which we don't have at the moment.
> * `Retry` and `RetryPolicy` - for retry strategies, but I believe we already have a possibility to configure retry strategy through Kernel instance, or it can be done on `HttpClient` level as well.
> 
> So, it looks like only `Diagnostics` property is missing, but again, in case of invalid subscription key, our current implementation should throw appropriate exception with error/status code and exception message. For better observability we can also configure `OpenAIClient` on our side with default `Diagnostics` values to provide more transparency for troubleshooting. We can also expose `List<string>` in Azure/OpenAI client constructors as optional parameter for `LoggedHeaderNames` property, but it will be abstract enough.
> 
> As per current implementation, we use plain .NET types to initialize Azure/OpenAI clients, like `string` endpoint, `string` key, `HttpClient`, `ILogger` etc. In case if we decide to replace Azure SDK with something else - it should not be a problem. But if we add Azure SDK as dependency in constructor, it will be harder to decouple it later.

The benefit here is the ability to inject this dependency with the configuration the client needs. There is an example in the PR description which shows where this is useful. 
I don't think we should try to allow a new way to configure the `OpenAIClientOptions` as we'd need to keep in sync with changes to the Azure OpenAI SDK.

**Review by shawncal — APPROVED** [review]

**Pull request merged**